src/IIIFPresentation/Repository/Paths/PathParser.cs

```diff
@@ -25,5 +25,40 @@ public static AssetId GetAssetIdFromNamedQueryCanvasId(this Canvas canvas, ILogg
             throw new FormatException($"Unable to extract AssetId from {canvas.Id}");
         }
     }
-}
 
+    public static string GetCanvasId(Models.API.Manifest.CanvasPainting canvasPainting, int customerId)
+    {
+        var canvasId = canvasPainting.CanvasId.ThrowIfNull(nameof(canvasPainting));
+
+        if (!Uri.IsWellFormedUriString(canvasId, UriKind.Absolute))
+        {
+            if (prohibitedCharacters.Any(pc => canvasId.Contains(pc)))
+            {
+                throw new ArgumentException($"canvas Id {canvasId} contains a prohibited character");
+            }
+
+            return canvasId;
+        }
+
+        var convertedCanvasId = new Uri(canvasId).PathAndQuery;
+        var startsWith = $"/{customerId}/canvases/";
+
+        if (!convertedCanvasId.StartsWith(startsWith) || convertedCanvasId.Length == startsWith.Length)
+            throw new ArgumentException($"Canvas Id {convertedCanvasId} is not valid");
+
+        var actualCanvasId =
+            convertedCanvasId.Substring(startsWith.Length, convertedCanvasId.Length - startsWith.Length);
+        if (prohibitedCharacters.Any(pc => actualCanvasId.Contains(pc)))
```

Comment:
Duplicate of lines 35-38 (but different as no braces). Create a method? Single place to manage this and the error message (e.g. if we switch from a black to a white list would involve updating in 1 place rather than 2).